abci/src/application.rs

```diff
@@ -115,66 +145,122 @@ pub trait RequestDispatcher: RefUnwindSafe {
     ///
     /// `RequestDispatcher` can indicate that it will no longer process new
     /// requests by returning `None` variant.
-    fn handle(&self, request: proto::abci::Request)
-        -> Result<Option<proto::abci::Response>, Error>;
+    fn handle(&self, request: abci::Request) -> Option<abci::Response>;
 }
 
 // Implement `RequestDispatcher` for all `Application`s.
 impl<A: Application + RefUnwindSafe> RequestDispatcher for A {
-    fn handle(
-        &self,
-        request: proto::abci::Request,
-    ) -> Result<Option<proto::abci::Response>, Error> {
-        tracing::trace!("Incoming request: {:?}", request);
-        let value = match request.value.unwrap() {
-            proto::abci::request::Value::Echo(req) => {
-                proto::abci::response::Value::Echo(self.echo(req))
-            },
-            proto::abci::request::Value::Flush(req) => {
-                proto::abci::response::Value::Flush(self.flush(req))
-            },
-            proto::abci::request::Value::Info(req) => {
-                proto::abci::response::Value::Info(self.info(req))
-            },
-            proto::abci::request::Value::InitChain(req) => {
-                proto::abci::response::Value::InitChain(self.init_chain(req))
-            },
-            proto::abci::request::Value::Query(req) => {
-                proto::abci::response::Value::Query(self.query(req))
-            },
-            proto::abci::request::Value::CheckTx(req) => {
-                proto::abci::response::Value::CheckTx(self.check_tx(req))
-            },
-            proto::abci::request::Value::OfferSnapshot(req) => {
-                proto::abci::response::Value::OfferSnapshot(self.offer_snapshot(req))
-            },
-            proto::abci::request::Value::LoadSnapshotChunk(req) => {
-                proto::abci::response::Value::LoadSnapshotChunk(self.load_snapshot_chunk(req))
-            },
-            proto::abci::request::Value::ApplySnapshotChunk(req) => {
-                proto::abci::response::Value::ApplySnapshotChunk(self.apply_snapshot_chunk(req))
-            },
-            proto::abci::request::Value::ListSnapshots(req) => {
-                proto::abci::response::Value::ListSnapshots(self.list_snapshots(req))
-            },
-            proto::abci::request::Value::PrepareProposal(req) => {
-                proto::abci::response::Value::PrepareProposal(self.prepare_proposal(req))
-            },
-            proto::abci::request::Value::ProcessProposal(req) => {
-                proto::abci::response::Value::ProcessProposal(self.process_proposal(req))
-            },
-            proto::abci::request::Value::FinalizeBlock(req) => {
-                proto::abci::response::Value::FinalizeBlock(self.finalize_block(req))
+    fn handle(&self, request: abci::Request) -> Option<abci::Response> {
+        tracing::trace!(?request, "received request");
+
+        let response: Result<response::Value, abci::ResponseException> = match request.value? {
```

Comment:
as I remember `handle` returns `None` in order to gracefully shutdown the server, if `request.value` is `None` this will happen instead of processing new connection